spacy/strings.pyx

```diff
@@ -14,236 +15,116 @@ from .symbols import NAMES as SYMBOLS_BY_INT
-        else:
-            raise TypeError(Errors.E017.format(value_type=type(string)))
-        return str_hash
+        for _, string in self.items():
+            yield string
+
+    def __reduce__(self):
+        strings = list(self)
+        return (StringStore, (strings,), None, None, None)
 
-    def __len__(self):
+    def __len__(self) -> int:
         """The number of strings in the store.
 
         RETURNS (int): The number of strings in the store.
         """
         return self.keys.size()
 
-    def __contains__(self, string_or_id not None):
-        """Check whether a string or ID is in the store.
+    def add(self, string: str) -> int:
+        """Add a string to the StringStore.
 
-        string_or_id (str or int): The string to check.
-        RETURNS (bool): Whether the store contains the string.
+        string (str): The string to add.
+        RETURNS (uint64): The string's hash value.
         """
-        cdef hash_t str_hash
-        if isinstance(string_or_id, str):
-            if len(string_or_id) == 0:
-                return True
-            elif string_or_id in SYMBOLS_BY_STR:
-                return True
-            str_hash = hash_string(string_or_id)
-        elif _try_coerce_to_hash(string_or_id, &str_hash):
-            pass
+        if not isinstance(string, str):
+            raise TypeError(Errors.E017.format(value_type=type(string)))
+
+        if string in SYMBOLS_BY_STR:
+            return SYMBOLS_BY_STR[string]
         else:
-            # TODO: Raise an error instead
-            return self._map.get(string_or_id) is not NULL
+            return self._intern_str(string)
 
-        if str_hash in SYMBOLS_BY_INT:
-            return True
+    def as_int(self, string_or_hash: Union[str, int]) -> str:
+        """If a hash value is passed as the input, return it as-is. If the input
+        is a string, return its corresponding hash.
+
+        string_or_hash (str, int): The string to hash or a hash value.
+        RETURNS (int): The hash of the string or the input hash value.
+        """
+        if isinstance(string_or_hash, int):
+            return string_or_hash
         else:
-            return self._map.get(str_hash) is not NULL
+            return get_string_id(string_or_hash)
 
-    def __iter__(self):
-        """Iterate over the strings in the store, in order.
+    def as_string(self, string_or_hash: Union[str, int]) -> str:
+        """If a string is passed as the input, return it as-is. If the input
+        is a hash value, return its corresponding string.
 
-        YIELDS (str): A string in the store.
+        string_or_hash (str, int): The hash value to lookup or a string.
+        RETURNS (str): The stored string or the input string.
+        """
+        if isinstance(string_or_hash, str):
+            return string_or_hash
+        else:
+            return self._get_interned_str(string_or_hash)
+
+    def items(self) -> Tuple[int, str]:
```

Comment:
Looking at the odd `Morphology` example above, do we maybe want to reverse this so that `for key in strings` and `for key, value in strings.items()` aren't swapped in terms of what looks like the key and what looks like the value?

Reply:
Yeah, fair point. Will swap the order.